libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
 				_functionCall.location(),
-				"Named arguments cannot be used for functions that take arbitrary parameters."
+				parameterNames.size() > argumentNames.size()
+				? "Some argument names are missing."
+				: "Too many arguments."
 			);
-		else if (parameterNames.size() > argumentNames.size())
-			m_errorReporter.typeError(_functionCall.location(), "Some argument names are missing.");
-		else if (parameterNames.size() < argumentNames.size())
-			m_errorReporter.typeError(_functionCall.location(), "Too many arguments.");
-		else
+			return;
+		}
+
+		// Check for duplicate argument names
 		{
-			// check duplicate names
 			bool duplication = false;
 			for (size_t i = 0; i < argumentNames.size(); i++)
 				for (size_t j = i + 1; j < argumentNames.size(); j++)
 					if (*argumentNames[i] == *argumentNames[j])
 					{
 						duplication = true;
-						m_errorReporter.typeError(arguments[i]->location(), "Duplicate named argument.");
+						m_errorReporter.typeError(
+							arguments[i]->location(),
+							"Duplicate named argument \"" + *argumentNames[i] + "\"."
+						);
 					}
+			if (duplication)
+				return;
+		}
+
+		// map parameter names to argument names
+		{
+			bool not_all_mapped = false;
+
+			for (size_t i = 0; i < param_arg_map_size; i++)
+			{
+				size_t j;
+				for (j = 0; j < argumentNames.size(); j++)
+					if (parameterNames[i] == *argumentNames[j])
+						break;
 
-			// check actual types
-			if (!duplication)
-				for (size_t i = 0; i < argumentNames.size(); i++)
+				if (j < argumentNames.size())
+					param_arg_map[i] = arguments[j].get();
+				else
 				{
-					bool found = false;
-					for (size_t j = 0; j < parameterNames.size(); j++)
-						if (parameterNames[j] == *argumentNames[i])
-						{
-							found = true;
-							// check type convertible
-							if (!type(*arguments[i])->isImplicitlyConvertibleTo(*parameterTypes[j]))
-								m_errorReporter.typeError(
-									arguments[i]->location(),
-									"Invalid type for argument in function call. "
-									"Invalid implicit conversion from " +
-									type(*arguments[i])->toString() +
-									" to " +
-									parameterTypes[i]->toString() +
-									" requested."
-								);
-							break;
-						}
-
-					if (!found)
-						m_errorReporter.typeError(
-							_functionCall.location(),
-							"Named argument \"" + *argumentNames[i] +  "\" does not match function declaration."
-						);
+					param_arg_map[i] = nullptr;
+					not_all_mapped = true;
+					m_errorReporter.typeError(
+						_functionCall.location(),
+						"Named argument \"" +
+						*argumentNames[i] +
+						"\" does not match function declaration."
+					);
 				}
+			}
+
+			if (not_all_mapped)
+				return;
+		}
+	}
+
+	// Check for compatible types between arguments and parameters
+	for (size_t i = 0; i < param_arg_map_size; ++i)
+		if (!type(*param_arg_map[i])->isImplicitlyConvertibleTo(*parameterTypes[i]))
+		{
+			solAssert(!!param_arg_map[i], "unmapped parameter");
```

Comment:
Isn't this too late for the assertion? ``param_arg_map[i]`` was already dereferenced in the ``if`` condition...

Reply:
Yes it is - fixed.